source/mir/ndslice/topology.d

```diff
@@ -3067,3 +3069,411 @@ unittest
+    assert(y.front == iota([3], 1, 4));
+}
+
+/// 3-dimensional slice support, N-dimensional also supported
+@safe @nogc pure nothrow
+unittest
+{
+    import mir.ndslice : iota, universal, strided, flattened, reshape;
+	
+    auto slice = iota(3, 4, 5);
+	
+    static immutable shape45 = [4, 5];
+    static immutable shape35 = [3, 5];
+    static immutable shape34 = [3, 4];
+    static immutable shape3 = [3];
+    static immutable shape4 = [4];
+    static immutable shape5 = [5];
+	
+    auto x = slice.byDim;
+    assert(x.shape == shape3);
+    assert(x.front.shape == shape45);
+    assert(x.front == iota([4, 5]));
+    x.popFront;
+    assert(x.front == iota([4, 5], (4 * 5)));
+	
+    auto y = slice.byDim!1;
+    assert(y.shape == shape4);
+    assert(y.front.shape == shape35);
+    int err;
+    assert(y.front == slice.universal.strided!1(4).reshape([3, -1], err));
+    y.popFront;
+    assert(y.front.front == iota([5], 5));
+	
+    auto z = slice.byDim!2;
+    assert(z.shape == shape5);
+    assert(z.front.shape == shape34);
+    assert(z.front == slice.universal.strided!2(5).reshape([3, -1], err));
+    z.popFront;
+    assert(z.front.front == iota([4], 1, 5));
+}
+
+// Ensure works on canonical
+@safe @nogc pure nothrow
+unittest
+{
+    import mir.ndslice : iota, canonical;
+	
+    auto slice = iota(3, 4).canonical;
+	
+    static immutable shape3 = [3];
+    static immutable shape4 = [4];
+	
+    auto x = slice.byDim;
+    assert(x.shape == shape3);
+    assert(x.front.shape == shape4);
+    assert(x.front == iota(4));
+    x.popFront;
+    assert(x.front == iota([4], 4));
+	
+    auto y = slice.byDim!1;
+    assert(y.shape == shape4);
+    assert(y.front.shape == shape3);
+    assert(y.front == iota([3], 0, 4));
+    y.popFront;
+    assert(y.front == iota([3], 1, 4));
+}
+
+// Ensure works on universal
+@safe @nogc pure nothrow
+unittest
+{
+    import mir.ndslice : iota, universal;
+	
+    auto slice = iota(3, 4).universal;
+	
+    static immutable shape3 = [3];
+    static immutable shape4 = [4];
+	
+    auto x = slice.byDim;
+    assert(x.shape == shape3);
+    assert(x.front.shape == shape4);
+    assert(x.front == iota(4));
+    x.popFront;
+    assert(x.front == iota([4], 4));
+	
+    auto y = slice.byDim!1;
+    assert(y.shape == shape4);
+    assert(y.front.shape == shape3);
+    assert(y.front == iota([3], 0, 4));
+    y.popFront;
+    assert(y.front == iota([3], 1, 4));
+}
+
+/++
+Convenience function to apply a function along a dimension of a slice.
+
+Combines $(LREF byDim) and $(LREF map) together for ease of use.
+
+Params:
+    f = function to feed to map
+    dim = dimension to perform function on
+    slice = input slice (may not be 1-dimensional slice)
+Returns:
+    1-dimensional slice of results of function applied to each dimension
+See_also:
+    $(LREF byDim),
+    $(LREF map),
+    $(HTTP https://stat.ethz.ch/R-manual/R-devel/library/base/html/apply.html, apply)
++/
+template apply(alias f, size_t dim = 0)
```

Comment:
Fair point on the naming, R's apply would really be more like `slice.byDim!dim.each!f`